.circleci/config.yml

```diff
@@ -1,79 +1,121 @@
+references:
+  #
+  # Environment
+  #
+  shared-environment: &shared-environment
+    CIRCLE_ARTIFACTS: /tmp/artifacts
+    CIRCLE_TEST_REPORTS: /tmp/test_results
+
+  #
+  # Image
+  #
+  docker-image: &docker-image
+    - image: circleci/node:10.10.0
+
+  #
+  # Cache management
+  #
+  restore-git-cache: &restore-git-cache
+    name: Restore git cache
+    keys:
+      - v2-git-{{ .Branch }}-{{ .Revision }}
+      - v2-git-{{ .Branch }}
+      - v2-git
+  save-git-cache: &save-git-cache
+    name: Save git cache
+    key: v2-git-{{ .Branch }}-{{ .Revision }}
+    paths:
+      - ".git"
+
+  restore-npm-cache: &restore-npm-cache
+    name: "Restore npm cache"
+    keys:
+       - v1-npmcache-{{ checksum ".nvmrc" }}-{{ checksum "npm-shrinkwrap.json" }}
+       - v1-npmcache-{{ checksum ".nvmrc" }}
+  save-npm-cache: &save-npm-cache
+    name: "Save npm cache"
+    key: v1-npmcache-{{ checksum ".nvmrc" }}-{{ checksum "npm-shrinkwrap.json" }}
+    paths:
+      - ~/.npm
+
+  #
+  # Artifact prep
+  #
+  setup-results-and-artifacts: &setup-results-and-artifacts
+    name: Create Directories for Results and Artifacts
+    command: mkdir -p                  \
+      $CIRCLE_ARTIFACTS                \
+      $CIRCLE_TEST_REPORTS/client      \
+      $CIRCLE_TEST_REPORTS/server
+
+
 version: 2
 jobs:
-  build:
-    parallelism: 6
+  #
+  # Shared setup required for all jobs
+  #
+  setup:
+    docker: *docker-image
     working_directory: ~/wp-calypso
-    docker:
-      - image: circleci/node:10.10.0-browsers
-        environment:
-          CIRCLE_ARTIFACTS: /tmp/artifacts
-          CIRCLE_TEST_REPORTS: /tmp/test_results
-
     steps:
-      - restore_cache:
-          name: Restore git cache
-          key: v1-20180524-git
-
+      - restore_cache: *restore-git-cache
       - checkout
-
-    # Update origin/master so we can diff against it later
+      # Update origin/master so we can diff against it later
       - run:
           name: Update master branch
           command: git fetch --force origin master
+      - save_cache: *save-git-cache
 
-      - save_cache:
-          name: Save git cache
-          key: v1-20180524-git
-          paths:
-            - ".git"
-
-      - run:
-          name: Create Directories for Results and Artifacts
-          command: |
-            mkdir -p $CIRCLE_ARTIFACTS
-            mkdir -p $CIRCLE_TEST_REPORTS
-
-      - restore_cache:
-          name: "Restoring npm cache"
-          keys:
-             - v1-npmcache-{{ checksum ".nvmrc" }}-{{ checksum "npm-shrinkwrap.json" }}
-             - v1-npmcache-{{ checksum ".nvmrc" }}
-
+      - restore_cache: *restore-npm-cache
       - run: npm ci
```

Comment:
does restoring the cache here simply mean that some packages won't have to download? I thought that performing `npm ci` would mean we can ignore caches because it would wipe out any existing packages

Reply:
In #27180 this changed to cache `node_modules` directly, skipping the npm installation on a cache hit. Cache matches checksums on both `.nvmrc` (should correspond to Node.js version in the container) and `npm-shrinkwrap.json`. From #27180:

> Add `node_modules` cache. Previously we had eliminated the `node_modules` cache because we were using `npm ci` which deletes the folder (#25431). Then, we maintained this behavior, but began to cache the npm cache so that `npm ci` would execute faster because the modules were already available locally (#25487). This approach restores the `node_modules` cache and _completely skips_ `npm ci` if there was a cache hit. This should be safe because the `node_modules` cache requires a checksum match on `.nvmrc` and `npm-shrinkwrap.json`. It's important to ensure that _all variables_ which might generate different `node_modules` structure are part of the cache key, e.g. running on different architectures.